lib/OpenQA/CLI/api.pm

```diff
@@ -26,28 +31,72 @@ sub command ($self, @args) {
       'd|data=s' => \$data,
       'f|form' => \my $form,
       'j|json' => \my $json,
+      'pre-script-hook=s' => \my $prehooksrc,
+      'post-script-hook=s' => \my $posthooksrc,
       'param-file=s' => \my @param_file,
       'r|retries=i' => \my $retries,
       'X|method=s' => \(my $method = 'GET');
 
     @args = $self->decode_args(@args);
     die $self->usage unless my $path = shift @args;
 
+    my $params;
+    my @data;
+    if ($prehooksrc) {
+
+	$self->_run_script ($_) for path($posthooksrc)->list->grep(qr/post_/)->each;
+	#my $hookdata = path($prehooksrc)->slurp;
+	#$data = ($hookdata);
+	#my $hookparams = decode_json($data);
+	#my $argsparams = $self->parse_params(\@args, \@param_file);
+	#$params = {%$hookparams, %$argsparams};
+    }
+    
     $data = path($data_file)->slurp if $data_file;
-    my @data = ($data);
-    my $params = $form ? decode_json($data) : $self->parse_params(\@args, \@param_file);
+    @data = ($data);
+    $params = $form ? decode_json($data) : $self->parse_params(\@args, \@param_file);
+    use Data::Dumper;
     @data = (form => $params) if keys %$params;
-
+    print Dumper(@data[1]);
+    if ($posthooksrc) {
+	use Data::Dumper;
+	# say for path('/home/sri/myapp')->list->each;
+	# run_script check exec permissions
+	$self->_run_script ($_, $data[1]) for path($posthooksrc)->list->grep(qr/post_/)->each;
```

Comment:
something important here, is to guarantee execution order, and document it 

```
for i in 1 9 10 19 29 21 01 09 20 100 200 101 110 109 199;do touch {pre_$i,post_$i}; done;
﻿foursixnine@pakhet test-sort % perl -Mojo -E 'say $_ for f->path->list->each'
/Users/foursixnine/tmp/test-sort/post_01
/Users/foursixnine/tmp/test-sort/post_09
/Users/foursixnine/tmp/test-sort/post_1
/Users/foursixnine/tmp/test-sort/post_10
/Users/foursixnine/tmp/test-sort/post_100
/Users/foursixnine/tmp/test-sort/post_101
/Users/foursixnine/tmp/test-sort/post_109
/Users/foursixnine/tmp/test-sort/post_110
/Users/foursixnine/tmp/test-sort/post_19
/Users/foursixnine/tmp/test-sort/post_199
/Users/foursixnine/tmp/test-sort/post_20
/Users/foursixnine/tmp/test-sort/post_200
/Users/foursixnine/tmp/test-sort/post_21
/Users/foursixnine/tmp/test-sort/post_29
/Users/foursixnine/tmp/test-sort/post_9
/Users/foursixnine/tmp/test-sort/pre_01
/Users/foursixnine/tmp/test-sort/pre_09
/Users/foursixnine/tmp/test-sort/pre_1
/Users/foursixnine/tmp/test-sort/pre_10
/Users/foursixnine/tmp/test-sort/pre_100
/Users/foursixnine/tmp/test-sort/pre_101
/Users/foursixnine/tmp/test-sort/pre_109
/Users/foursixnine/tmp/test-sort/pre_110
/Users/foursixnine/tmp/test-sort/pre_19
/Users/foursixnine/tmp/test-sort/pre_199
/Users/foursixnine/tmp/test-sort/pre_20
/Users/foursixnine/tmp/test-sort/pre_200
/Users/foursixnine/tmp/test-sort/pre_21
/Users/foursixnine/tmp/test-sort/pre_29
/Users/foursixnine/tmp/test-sort/pre_9
```

Reply:
I'm also wondering whether it makes sense to run multiple scripts. I guess I would not even have separate pre/post scripts. Or can you explain what the use case of those will be?

Reply:
> Or can you explain what the use case of those will be?

pre_hook: Add very_expensive_testsuite if repodata contains $_SPECIAL_STUFF
post_hook: turn on very_expensive_machine if repodata contains $_SPECIAL_STUFF

one instance of such actions _could be_ spinning up a Grace Hopper machine, or making an api call to a third service

Reply:
I would expect that the scripts are atomic and independent. I can think of some implications. What happens if one script overwrites previous variables for example.

Reply:
regarding the order: `list()` does a sort, so the order is guaranteed. it's alphanumeric, so if you want to order by number it probably makes sense to have the same number of digits, e.g. `post_001`, `post_002` etc.